tool/src/c2/ty.rs

```diff
@@ -0,0 +1,275 @@
+    let header_path = format!("{name}.h");
+    let mut header = Header::new(name.to_owned());
+    let mut context = TyGenContext::new(cx, &mut header);
+    context.gen_result(name, ty);
+    cx.files.add_file(header_path, header.to_string());
+}
+
+pub type ResultType<'tcx> = (Option<&'tcx hir::OutType>, Option<&'tcx hir::OutType>);
+
+/// Context for generating a particular type's header
+pub struct TyGenContext<'cx, 'tcx, 'header> {
+    cx: &'cx CContext<'tcx>,
+    header: &'header mut Header,
+}
+
+impl<'cx, 'tcx: 'cx, 'header> TyGenContext<'cx, 'tcx, 'header> {
+    pub fn new(cx: &'cx CContext<'tcx>, header: &'header mut Header) -> Self {
+        TyGenContext { cx, header }
+    }
+
+    pub fn gen_enum_def(&mut self, def: &'tcx hir::EnumDef, id: TypeId) {
+        // Enums can't be forward-declared in C, but we do want enums to have methods,
+        // which may require additional #includes leading to potential cycles.
+        // To handle this, we make a separate header file called Foo_enum.h, that contains
+        // *just* the enum. It is included from Foo.h, and external users should not be importing
+        // it directly. (We can potentially add a #define guard that makes this actually private, if needed)
+        let header_name = &self.header.identifier;
+        let enum_header_name = format!("{header_name}_enum");
+        self.header.includes.insert(enum_header_name.to_string());
+        let enum_header_path = format!("{enum_header_name}.h");
+        let mut enum_header = Header::new(enum_header_name);
+
+        let ty_name = self.cx.fmt_type_name(id);
+        enum_header.body += &format!("typedef enum {ty_name} {{\n");
+        for variant in def.variants.iter() {
+            let variant_name = self.cx.fmt_enum_variant(variant);
+            let discriminant = variant.discriminant;
+            enum_header.body += &format!("\t{ty_name}_{variant_name} = {discriminant},\n")
+        }
+        enum_header.body += &format!("}} {ty_name};\n");
+
+        self.cx
+            .files
+            .add_file(enum_header_path, enum_header.to_string());
+    }
+
+    pub fn gen_opaque_def(&mut self, _def: &'tcx hir::OpaqueDef, id: TypeId) {
+        let ty_name = self.cx.fmt_type_name(id);
+        self.header.body += &format!("typedef struct {ty_name} {ty_name};\n");
+    }
+
+    pub fn gen_struct_def<P: TyPosition>(&mut self, def: &'tcx hir::StructDef<P>, id: TypeId) {
+        let ty_name = self.cx.fmt_type_name(id);
+        self.header.body += &format!("typedef struct {ty_name} {{\n");
+        for field in def.fields.iter() {
+            let decls = self.gen_ty_decl(&field.ty, field.name.as_str(), true);
+            for (decl_ty, decl_name) in decls {
+                self.header.body += &format!("\t{decl_ty} {decl_name};\n")
+            }
+        }
+        // reborrow to avoid borrowing across mutation
+        self.header.body += &format!("}} {ty_name};\n");
```

Comment:
**nit**: write this all to a `String` first using `write!()` to avoid repeated `+= &format!(...)` calls, and then append it to `self.header.body` at the end.